File: app/src/main/java/com/mkulesh/onpc/iscp/messages/TimeInfoMsg.java
/*
 * Enhanced Music Controller
 * Copyright (C) 2018-2023 by Mikhail Kulesh
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU
 * General Public License as published by the Free Software Foundation, either version 3 of the License,
 * or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
 * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details. You should have received a copy of the GNU General
 * Public License along with this program.
 */

package com.mkulesh.onpc.iscp.messages;

import android.annotation.SuppressLint;

import com.mkulesh.onpc.iscp.EISCPMessage;
import com.mkulesh.onpc.iscp.ISCPMessage;
import com.mkulesh.onpc.utils.Utils;

import java.util.Map;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

/*
 * NET/USB Time Info
 */
public class TimeInfoMsg extends ISCPMessage
{
    public final static String CODE = "NTM";
    public final static String INVALID_TIME = "--:--:--";

    /*
     * (Elapsed time/Track Time Max 99:59:59. If time is unknown, this response is --:--)
     */
    private final String currentTime;
    private final String maxTime;

    TimeInfoMsg(EISCPMessage raw) throws Exception
    {
        super(raw);
        final String[] pars = data.split(PAR_SEP);
        if (pars.length != 2)
        {
            throw new Exception("Can not find parameter split character in message " + raw);
        }
        currentTime = pars[0];
        maxTime = pars[1];
    }

    TimeInfoMsg(String currentTime, String maxTime)
    {
        super(0, null);
        this.currentTime = currentTime;
        this.maxTime = maxTime;
    }

    public String getCurrentTime()
    {
        return (currentTime == null || currentTime.isEmpty()) ? INVALID_TIME : currentTime;
    }

    public String getMaxTime()
    {
        return (maxTime == null || maxTime.isEmpty()) ? INVALID_TIME : maxTime;
    }

    @NonNull
    @Override
    public String toString()
    {
        return CODE + "[" + currentTime + "; " + maxTime + "]";
    }

    /*
     * Denon control protocol
     * - Player Now Playing Progress
     * {
     * "heos": {
     *     "command": " event/player_now_playing_progress",
     *     "message": "pid=player_id&cur_pos=position_ms&duration=duration_ms"
     *     }
     * }
     */
    private final static String HEOS_COMMAND = "event/player_now_playing_progress";

    @Nullable
    @SuppressLint("SimpleDateFormat")
    public static TimeInfoMsg processHeosMessage(@NonNull final String command, @NonNull final Map<String, String> tokens)
    {
        if (HEOS_COMMAND.equals(command))
        {
            final String curPosStr = tokens.get("cur_pos");
            final String durationStr = tokens.get("duration");
            if (curPosStr != null && durationStr != null)
            {
                return new TimeInfoMsg(
                        Utils.millisToTime(Integer.parseInt(curPosStr)),
                        Utils.millisToTime(Integer.parseInt(durationStr)));
            }
        }
        return null;
    }
}

File: lib/iscp/messages/TimeInfoMsg.dart
/*
 * Enhanced Music Controller
 * Copyright (C) 2019-2023 by Mikhail Kulesh
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU
 * General Public License as published by the Free Software Foundation, either version 3 of the License,
 * or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
 * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details. You should have received a copy of the GNU General
 * Public License along with this program.
 */

import 'package:sprintf/sprintf.dart';

import "../DcpHeosMessage.dart";
import "../EISCPMessage.dart";
import "../ISCPMessage.dart";

/*
 * NET/USB Time Info
 */
class TimeInfoMsg extends ISCPMessage
{
    static const String CODE = "NTM";
    static const String INVALID_TIME = "--:--:--";

    /*
     * (Elapsed time/Track Time Max 99:59:59. If time is unknown, this response is --:--)
     */
    late String _currentTime, _maxTime;

    TimeInfoMsg(EISCPMessage raw) : super(CODE, raw)
    {
        final List<String> pars = getData.split(ISCPMessage.PAR_SEP);
        if (pars.length != 2)
        {
            throw Exception("Can not find parameter split character in message " + raw.toString());
        }
        _currentTime = ISCPMessage.nonNullString(pars[0]);
        _maxTime = ISCPMessage.nonNullString(pars[1]);
    }

    TimeInfoMsg._dcp(String currentTime, String maxTime) : super.output(CODE, "")
    {
        _currentTime = currentTime;
        _maxTime = maxTime;
    }

    String get getCurrentTime
    => _currentTime.isEmpty ? INVALID_TIME : _currentTime;

    String get getMaxTime
    => _maxTime.isEmpty ? INVALID_TIME : _maxTime;

    @override
    String toString()
    => super.toString() + "[CURR=" + getCurrentTime + "; MAX=" + getMaxTime + "]";

    /*
     * Denon control protocol
     * - Player Now Playing Progress
     * {
     * "heos": {
     *     "command": " event/player_now_playing_progress",
     *     "message": "pid=player_id&cur_pos=position_ms&duration=duration_ms"
     *     }
     * }
     */
    static const String _HEOS_COMMAND = "event/player_now_playing_progress";

    static TimeInfoMsg? processHeosMessage(DcpHeosMessage jsonMsg)
    {
        if (_HEOS_COMMAND == jsonMsg.command)
        {
            final String? curPosStr = jsonMsg.message["cur_pos"];
            final String? durationStr = jsonMsg.message["duration"];
            if (curPosStr != null && durationStr != null)
            {
                final int? curPos = int.tryParse(curPosStr);
                final int? duration = int.tryParse(durationStr);
                if (curPos != null && duration != null)
                {
                    return TimeInfoMsg._dcp(_millisToTime(curPos),
                        duration == 0 ? INVALID_TIME : _millisToTime(duration));
                }
            }
        }
        return null;
    }

    static String _millisToTime(int millis)
    {
        final int inpSec = (millis / 1000).floor();
        final int hours = (inpSec / 3600).floor();
        final int minutes = ((inpSec - hours * 3600) / 60).floor();
        final int seconds = inpSec - hours * 3600 - minutes * 60;
        return sprintf("%02d:%02d:%02d", [ hours, minutes, seconds ]);
    }
}
